src/dinit.cc

```diff
@@ -696,6 +700,34 @@ int dinit_main(int argc, char **argv)
     close_control_socket();
     
     if (am_system_mgr) {
+        if (shutdown_type == shutdown_type_t::SOFTREBOOT) {
+            sync(); // Sync to minimise data loss if user elects to power off / hard reset
+
+            try {
+                char **argv_exec = new char*[argc + 1];
+                int j = 0;
+                for (int i = 0; i < argc; i++) {
+                    // strip out kernel cmdline: single
+                    if (strcmp(argv[i], "single") != 0) {
```

Comment:
Sorry for not being able to look at this earlier. As I mentioned, I needed to find a time when I wasn't busy to figure out the right solution. Having thought about it for a bit, I think I have some better idea now.

This current solution, filtering just the "single" argument, isn't sufficient because there are a number of ways users can pass additional service names on the command line argument (via the bootloader):

- `single` (on Linux) which is covered by your current changes
- `-t servicename` for any service name or `--service servicename` for any service name
- `-m servicename1 servicename2 servicename...` for any service names
- `-c servicename1 servicename2 servicename...` for any service names (container mode)

I think the best option is to, when processing the command line arguments initially, construct an array (probably a `vector<boolean>` actually) which keeps track of which arguments should be filtered out on re-exec. Then we don't need to duplicate the argument processing logic; it's all in one place.

Reply:
Alternatively, update the documentation to be very clear about what a "soft reboot" actually does. It seems reasonable to me that it will just restart with the exact same arguments, but in that case the documentation needs to be clear and warn that the original boot-time options will be re-applied.

Reply:
I just tested this on systemd and found that the `single` parameter is still preserved after soft-reboot, so let's just use `argv` directly to be consistent with it in this behavior.

Reply:
> Usecase: this is an extremely quick way to reset userspace fully when updating image based systems, without going through a full hardware/firmware/boot loader/kernel reset. It minimizes "grayout time" for OS updates. (In particular when combined with kernel live patching)

From the original PR for systemd: https://github.com/systemd/systemd/pull/27435